Improve the dump-arrays performance on Windows

hjyamauchi opened the pull request:
Use the HeapWalk API for heap iteration instead of the
Heap32First/Next API, which was known to be slow. Since HeapWalk only
works locally, it requires using a remote thread and a DLL.

hjyamauchi committed:
Improve the dump-arrays performance on Windows

Use the HeapWalk API for heap iteration instead of the
Heap32First/Next API, which was known to be slow. Since HeapWalk only
works locally, it requires using a remote thread and a DLL.

compnerd commented on tools/swift-inspect/Sources/SwiftInspectHeapWalk/SwiftInspectHeapWalk.cpp:
```diff
@@ -0,0 +1,184 @@
+
+#define BUF_SIZE 512
+#define BUF_NUM_ENTRIES (BUF_SIZE / sizeof(HeapEntry))
+#define SHARED_MEM_NAME "Local\\SwiftInspectFileMapping"
+#define READ_EVENT_NAME "Local\\SwiftInspectReadEvent"
+#define WRITE_EVENT_NAME "Local\\SwiftInspectWriteEvent"
```
Do we need to worry about collisions in case of multiple running entries?  Perhaps we should use `.{ProcessID}` to disambiguate?

compnerd commented on tools/swift-inspect/Sources/SwiftInspectHeapWalk/SwiftInspectHeapWalk.cpp:
```diff
@@ -0,0 +1,184 @@
+    return 1;
+  }
+  HeapEntry *buf = (HeapEntry *)
+    MapViewOfFile(hMapFile,
```
We should follow LLVM style, perhaps `clang-format` the file?

compnerd commented on tools/swift-inspect/Sources/SwiftInspectHeapWalk/SwiftInspectHeapWalk.cpp:
```diff
@@ -0,0 +1,184 @@
+    printf("OpenFileMapping failed\n");
+    return 1;
+  }
+  HeapEntry *buf = (HeapEntry *)
```
Since the file is marked as C++ rather than C, please use `reinterpret_cast` rather than the C-style cast.

compnerd commented on tools/swift-inspect/Sources/SwiftInspectHeapWalk/SwiftInspectHeapWalk.cpp:
```diff
@@ -0,0 +1,184 @@
+                  0,
+                  BUF_SIZE);
+  if (buf == NULL) {
+    printf("MapViewOfFile failed\n");
```
How can we see the `printf` output?  Would this potentially be printed into the output of the client?  Perhaps we should use `OutputDebugString`?

compnerd commented on tools/swift-inspect/Sources/SwiftInspectHeapWalk/SwiftInspectHeapWalk.cpp:
```diff
@@ -0,0 +1,184 @@
+                  BUF_SIZE);
+  if (buf == NULL) {
+    printf("MapViewOfFile failed\n");
+    CloseHandle(hMapFile);
```
Since this is C++ code, would you mind adding in a `scoped_handle`?  (See https://github.com/apple/swift-installer-scripts/commit/927c674fb3278d2e61a63e3d7c8e7efd0ff3aa2e#diff-baf3457e394df6903a81ffffaaf57a045306145a255ce7c5a5b34ba01e7f7d56L25-L33 if you would like a reference)

compnerd commented on tools/swift-inspect/Sources/SwiftInspectHeapWalk/SwiftInspectHeapWalk.cpp:
```diff
@@ -0,0 +1,184 @@
+    CloseHandle(hMapFile);
+    return 1;
+  }
+  memset(buf, 0, BUF_SIZE);
```
When in C++, use `std::memset` over `memset`

compnerd commented on tools/swift-inspect/Sources/SwiftInspectHeapWalk/SwiftInspectHeapWalk.cpp:
```diff
@@ -0,0 +1,184 @@
+    CloseHandle(hMapFile);
+    UnmapViewOfFile(buf);
+    CloseHandle(writeEvent);
+    CloseHandle(readEvent);
```
Use of `scoped_handle` should take care of three of these and leave only `MapViewOfFile`.  I wonder if we can use `std::unique_ptr` with a custom deleter for that.

compnerd commented on tools/swift-inspect/Sources/SwiftInspectHeapWalk/SwiftInspectHeapWalk.cpp:
```diff
@@ -0,0 +1,184 @@
+      break;
+    }
+    heapCount = actualHeapCount;
+    free(heaps);
+    heaps = (HANDLE*)malloc(heapCount * sizeof(HANDLE));
```
We are in C++, so prefer to use `delete[]` and `new[]`.

compnerd commented on tools/swift-inspect/Sources/SwiftInspectHeapWalk/SwiftInspectHeapWalk.cpp:
```diff
@@ -0,0 +1,184 @@
+  }
+
+  auto cleanup = [&] {
+    free(heaps);
```
Custom deleter with `std::unique_ptr` should avoid this right?

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,37 +236,263 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+    let sharedMemoryName = "Local\\SwiftInspectFileMapping"
+    let readEventName = "Local\\SwiftInspectReadEvent"
+    let writeEventName = "Local\\SwiftInspectWriteEvent"
+    let waitTimeoutMs = DWORD(30000)
```
How about extracting these constants from the C++ side into a C-only header and then using a module to share the constants across the C++ and Swift sides?

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,37 +236,263 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+    defer { CloseHandle(hWriteEvent) }
+
+    // The path to the dll assuming it's in the same directory as swift-inspect.
+    let dllPath: String = Bundle.main.bundlePath + "\\SwiftInspectHeapWalk.dll"
```
Hmm, if we use the bundle, that makes deploying this a bit more difficult as that has requirements on the layout.

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,37 +236,263 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+        return
+      }
+      // Allocate memory and write dllPath in the remote process
+      dllPathUtf16Len = UInt(wcslen($0) * 2)  // *2 for wide chars
```
I think that I would prefer using `MemoryLayout` rather than the hardcoded 2.

hjyamauchi commented:
The latest revision should address all the comments.

tristanlabelle commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,37 +237,266 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+    // Load the dll and start the heap walk
+    let hookAddr = unsafeBitCast(GetProcAddress(hKernel32, "LoadLibraryW"),
+                                 to: LPTHREAD_START_ROUTINE.self)
+    let hThread: HANDLE = CreateRemoteThread(self.process, nil, 0, hookAddr,
```
Mind blown that we can use `LoadLibraryW` as the entry point!

compnerd replied:
Yeah, that was the magic trick that I had in mind for this.  The signature is proper, and `DllMain` with the flags allows us to inject execution nicely.

tristanlabelle commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,37 +237,266 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+    defer { CloseHandle(hWriteEvent) }
+
+    // The path to the dll assuming it's in the same directory as swift-inspect.
+    let swiftInspectPath = ProcessInfo.processInfo.arguments[0]
```
Seems like this block could easily be extracted into a function.

hjyamauchi replied:
Done

tristanlabelle commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,37 +237,266 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+    }
+
+    // Unload the dll from the remote process
+    let unloadHookAddr = unsafeBitCast(GetProcAddress(hKernel32, "FreeLibrary"),
```
This block also seems easily extractable to a func.

tristanlabelle reviewed (commented):
Mind-blowing! It'd be great if the code could be broken up into a few functions so that the high-level logic can be read more easily.

hjyamauchi commented:
@compnerd Can you take another look?

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,37 +237,286 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+    var dllPathRemote = UnsafeMutableRawPointer(bitPattern: 0)
+    var dllPathUtf16Len: UInt = 0
+    var dllPathWritten = false
+    dllPath.withCString(encodedAs: UTF16.self) {
```
This is unsafe.  We have a path limitation to 261 characters.  We absolutely need to do this inside the closure to ensure that we have a NT-style path ...


```swift
URL(fileURLWithPath: swiftInspectPath)
        .deletingLastPathComponent()
        .appendingPathComponent("SwiftInspectClient.dll")
        .withUnsafeFileSystemRepresentation {
  #"\\?\\#(String(decodingCString: $0!, as: UTF16.self))"#.withCString(encodedAs: UTF16.self) {
    let faAttributes: WIN32_FILE_ATTRIBUTE_DATA = .init()
    guard GetFileAttributesExW($0, GetFileExInfoStandard, &faAttributes),
          faAttributes.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT == 0 else {
      return nil
    }

    let szLength = SIZE_T(_wcslen($0) + 1)
    guard let allocation = VirtualAllocEx(self.process, nil, szLength,
                                          DWORD(MEM_COMIT), DWORD(PAGE_READWRITE)) else {
      return nil
    }

    if !WriteProcessMemory(self.process, allocation, $0, szLength, nil) {
      VirtualFreeEx(self.process, allocation, 0, DWORD(MEM_RELEASE)
      return nil
    }

    return allocation
  }
}
```

hjyamauchi replied:
Done. Can you elaborate? I understand there's a path limitation. What's a NT-style path? And how does using a closure ensure it? Or what did the original code do wrong?

compnerd replied:
The `\\?\` prefix creates a NT-style path.  It is a path that bypasses the Win32 layer and goes into the kernel namespace, which allows it to bypass the 261 character limitation.  The closure doesn't do that, its the interior string construction which does that.  The problem is that the path must be normalised first as the normalization is something that the Win32 layer normally handles.  Due to the memory handling in Foundation, we need to use a closure (`withUnsafeFileSystemRepresentation`).  That mostly normalizes the path, and then we can create the long path.

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,37 +237,286 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+
+  private func unloadDllAndPathRemote(
+      dwProcessId: DWORD, hKernel32: HMODULE!,
+      dllPathRemote: UnsafeMutableRawPointer) -> Bool {
```
Was this formatted by `swift-format`?

hjyamauchi replied:
Is there a swift-format config file that's used for the existing code? If I run it with its default config, it'd change the lines that aren't touched by this PR. Any preference?

compnerd replied:
I don't think that I had saved it ... the default sounds like a good idea though.  I think that there may be a few places where it will break before the first parameter which is probably the only divergence the existing code has.

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,37 +237,286 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+    let hModuleSnapshot: HANDLE =
+        CreateToolhelp32Snapshot(DWORD(TH32CS_SNAPMODULE), dwProcessId)
+    if hModuleSnapshot == INVALID_HANDLE_VALUE {
+      print("CreateToolhelp32Snapshot failed")
```
This doesn't give us much context on what happened, can we get extended information?

hjyamauchi replied:
Done

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,37 +237,286 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+    var entry: MODULEENTRY32W = MODULEENTRY32W()
+    entry.dwSize = DWORD(MemoryLayout<MODULEENTRY32W>.size)
+    if !Module32FirstW(hModuleSnapshot, &entry) {
+      print("Module32FirstW failed")
```
Similar

hjyamauchi replied:
Done

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,37 +237,286 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+    }
+    var entry: MODULEENTRY32W = MODULEENTRY32W()
+    entry.dwSize = DWORD(MemoryLayout<MODULEENTRY32W>.size)
+    if !Module32FirstW(hModuleSnapshot, &entry) {
```
Nit: perhaps convert to `guard`?

hjyamauchi replied:
Done

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,37 +237,286 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+      let module: String = withUnsafePointer(to: entry.szModule) {
+        $0.withMemoryRebound(
+          to: WCHAR.self,
+          capacity: MemoryLayout.size(ofValue: $0) / MemoryLayout<WCHAR>.size) {
```
This seems wrong?  This is going to be `sizeof(void *)` right?  Since you have an `UnsafeRawPointer` right?

hjyamauchi replied:
??? This looks to me like casting a char[] to wchar[] as szModule is a char[] that stores the name of the module. This follows what init? does above around line 171 above.

compnerd replied:
I think that I had misread the value of `$0`.  Sorry!

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,37 +237,286 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+    entry.dwSize = DWORD(MemoryLayout<MODULEENTRY32W>.size)
+    if !Module32FirstW(hModuleSnapshot, &entry) {
+      print("Module32FirstW failed")
+      return false
```
There is a leaked handle here.

hjyamauchi replied:
Done

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,37 +237,286 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+      print("CreateToolhelp32Snapshot failed")
+      return false
+    }
+    var entry: MODULEENTRY32W = MODULEENTRY32W()
```
Insert a:

```swift
defer { CloseHandle(hModuleSnapshot) }
```

to avoid a leaked handle (mentioned below)

hjyamauchi replied:
Done

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,37 +237,286 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+      return false
+    }
+    defer { CloseHandle(hUnloadThread) }
+    let unload_wait = WaitForSingleObject(hUnloadThread, DWORD(WAIT_TIMEOUT_MS))
+    if unload_wait != WAIT_OBJECT_0 {
```
I'd use `guard` and remove the variable.

hjyamauchi replied:
I'd like to keep it as it'd indicate the reason for the return

compnerd commented on tools/swift-inspect/Sources/SwiftInspectClientShims/module.modulemap:
```diff
@@ -0,0 +1,4 @@
+module SwiftInspectClientShims {
```
These aren't really shims, they are part of the interface itself.

hjyamauchi replied:
Done

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,37 +237,286 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+      print("CreateFileMapping failed \(GetLastError())")
+      return
+    }
+    if GetLastError() == ERROR_ALREADY_EXISTS {
+      print("FileMapping in use. Try again in a while.")
+      CloseHandle(hMapFile)
+      return
+    }
```
Does this matter if `hMapFile` is not-null?

hjyamauchi replied:
Doesn't matter as we suffix the pid in the name. Removed.

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,37 +237,286 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+      CloseHandle(hMapFile)
+      return
+    }
+    defer { CloseHandle(hMapFile) }
```
Should this be hoisted over the `GetLastError`?

hjyamauchi replied:
Done

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,37 +237,286 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+      return
+    }
+    defer { CloseHandle(hMapFile) }
+    let FILE_MAP_ALL_ACCESS = STANDARD_RIGHTS_REQUIRED | SECTION_QUERY |
```
I assume that this mirrors `FILE_MAP_ALL_ACCESS` in Windows.h?  If so, can we move this to a `WinSDK+Extensions.swift` please (with `internal` access)?

hjyamauchi replied:
Done

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,37 +237,286 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+                                    DWORD(FILE_MAP_ALL_ACCESS),
+                                    0,
+                                    0,
+                                    SIZE_T(bufSize));
```
Is this formatted via `swift-format`?

hjyamauchi replied:
see above

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,37 +237,286 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+    }
+    defer { UnmapViewOfFile(buf) }
+
+    // Set up the event objects
+    let hReadEvent: HANDLE = CreateEventA(LPSECURITY_ATTRIBUTES(bitPattern: 0),
+                                          false, // Auto-reset
+                                          false, // Initial state is nonsignaled
+                                          readEventName)
+    if hReadEvent == HANDLE(bitPattern: 0) {
+      print("CreateEvent failed \(GetLastError())")
+      return
+    }
+    defer { CloseHandle(hReadEvent) }
+    let hWriteEvent: HANDLE = CreateEventA(LPSECURITY_ATTRIBUTES(bitPattern: 0),
+                                           false, // Auto-reset
+                                           false, // Initial state is nonsignaled
+                                           writeEventName)
+    if hWriteEvent == HANDLE(bitPattern: 0) {
+      print("CreateEvent failed \(GetLastError())")
       return
     }
+    defer { CloseHandle(hWriteEvent) }
```
Might be nice to have a `CreateEventPair(_:) -> (HANDLE, HANDLE)` instead to create the named event pair.

hjyamauchi replied:
Done

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,37 +237,286 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+    }
+
+    // Load the dll and start the heap walk
+    let hookAddr = unsafeBitCast(GetProcAddress(hKernel32, "LoadLibraryW"),
+                                 to: LPTHREAD_START_ROUTINE.self)
```
This seems unnecessary.  It resolves to the local function `LoadLibraryW`, so why do the dynamic lookup?

hjyamauchi replied:
Casting the local function pointer `LoadLibraryW` to  `UnsafeMutableRawPointer` or `LPTHREAD_START_ROUTINE` doesn't seem to work with `Swift/Builtin.swift:95: Fatal error: Can't unsafeBitCast between types of different sizes`. How?

compnerd replied:
```swift
let lpfnLoadLibraryW: @convention(c) (LPCWSTR?) -> HMODULE? = LoadLibraryW
_ = unsafeBitCast(lpfnLoadLibraryW, to: LPTHREAD_START_ROUTINE.self)
```

I imagine you can even combine the two lines.

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,37 +237,286 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+        return
+      }
+
+      let entryCount = bufSize/MemoryLayout<(UInt, UInt)>.size
```
This is unsafe. You cannot marshal `(UInt, UInt)` over the wire as it is a Swift type and has no analogue in C.

hjyamauchi replied:
Done

hjyamauchi commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,37 +237,286 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+    let swiftInspectPath = ProcessInfo.processInfo.arguments[0]
+    let dllPath: String =
+        URL(fileURLWithPath: swiftInspectPath).deletingLastPathComponent().path +
+        "\\SwiftInspectClient.dll"
```
Done

hjyamauchi commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,37 +237,286 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+  }
+
+  private func allocateDllPathRemote() -> UnsafeMutableRawPointer? {
+    // The path to the dll assuming it's in the same directory as swift-inspect.
+    let swiftInspectPath = ProcessInfo.processInfo.arguments[0]
```
Agreed. If we need it, any reference on how to do that?

compnerd replied:
https://learn.microsoft.com/en-us/windows/win32/menurc/using-resources is a good starting point for that.  It explains how to locate and add the resources to the executable.

CodaFi commented:
I’m no expert in win32, but this looks pretty good. I guess my gut reaction reading all of this is that anything that touches HeapLock needs documentation about avoiding further dynamic allocations until the lock is dropped. It’s so easy to re-enter the allocator holding this thing.

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,38 +237,277 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+    let (hReadEvent, hWriteEvent) = createEventPair(dwProcessId)
+    if hReadEvent == HANDLE(bitPattern: 0) {
+      print("CreateEvent failed \(GetLastError())")
+      return
```
Can this not leak `hWriteEvent`?  Can we sink the error checks inside the `createEventPair`?

hjyamauchi replied:
Done

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,38 +237,277 @@ internal final class WindowsRemoteProcess: RemoteProcess {
-        if entry.dwFlags & DWORD(LF32_FREE) == DWORD(LF32_FREE) { continue }
-        body(swift_addr_t(entry.dwAddress), UInt64(entry.dwBlockSize))
-      } while Heap32Next(&entry)
+      let entryCount = bufSize/MemoryLayout<HeapEntry>.size
```
```suggestion
      let entryCount = bufSize / MemoryLayout<HeapEntry>.size
```

hjyamauchi replied:
Done

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,38 +237,277 @@ internal final class WindowsRemoteProcess: RemoteProcess {
-        body(swift_addr_t(entry.dwAddress), UInt64(entry.dwBlockSize))
-      } while Heap32Next(&entry)
+      let entryCount = bufSize/MemoryLayout<HeapEntry>.size
+      let buffer = buf.bindMemory(to: HeapEntry.self, capacity: 1)
+      let entry = buffer.pointee
+      if entry.Address == UInt.max {
+        // The buffer containing zero entries, indicated by the first entry
+        // contains -1, means, we are done. Break out of loop.
+        if !SetEvent(hWriteEvent) {
+          print("SetEvent failed \(GetLastError())")
+          return
+        }
+        break
+      }
+      for i in 0..<entryCount {
+        let entry = buffer.advanced(by: i).pointee
+        if entry.Address == 0 {
+          // Done. Break out of loop.
+          break
+        }
+        body(swift_addr_t(entry.Address), UInt64(entry.Size))
+      }
```
I think that this could be a bit nicer as:

```swift
for entry in UnsafeMutableBufferPointer(start: buf.bindMemory(to: HeapEntry.self,
                                                              capacity: entryCount),
                                        count: entryCount) {
  if entry.Address == UInt.max {
    if !SetEvent(hWriteEvent) {
      print("SetEvent failed: \(GetLastError())")
      return
    }
    break
  }
  if entry.Address == 0 { break }
  body(swift_addr_t(entry.Address), UInt64(entry.Size))
}
```

This avoids the `advanced(by:)` dance.

hjyamauchi replied:
Done. Is there a way to check the iteration index in this for-in loop to have the UInt.max check is for the first entry only like `&entry == buf`?

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,38 +237,277 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+        .deletingLastPathComponent()
+        .appendingPathComponent("SwiftInspectClient.dll")
+        .withUnsafeFileSystemRepresentation {
+      #"\\?\\#(String(decodingCString: unsafeBitCast($0!, to: UnsafePointer<UInt8>.self), as: UTF8.self))"#
```
Why is the `unsafeBitCast` needed here?

hjyamauchi replied:
A cast from `UnsafePointer<Int8>` to `UnsafePointer<UInt8>` needed here

hjyamauchi replied:
That doesn't seem to work. 

```
S:\SourceCache\swift\tools\swift-inspect\Sources\swift-inspect\WindowsRemoteProcess.swift:421:45: error: value of type 'UnsafePointer<Int8>?' has no member 'baseAddress'
        #"\\?\\#(String(decodingCString: $0.baseAddress, as: UTF8.self))"#
                                         ~~ ^~~~~~~~~~~
```
```
S:\SourceCache\swift\tools\swift-inspect\Sources\swift-inspect\WindowsRemoteProcess.swift:421:46: error: value of type 'UnsafePointer<Int8>' has no member 'baseAddress'
        #"\\?\\#(String(decodingCString: $0!.baseAddress, as: UTF8.self))"#
```

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,38 +237,277 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+
+    // Unload the dll from the remote process
+    let unloadHookAddr = unsafeBitCast(GetProcAddress(hKernel32, "FreeLibrary"),
+                                       to: LPTHREAD_START_ROUTINE.self)
```
Similar to above for `LoadLibraryW`, we can probably avoid the dynamic resolution of `FreeLibrary`.

hjyamauchi replied:
```
    let lpfnFreeLibrary:  (HMODULE?) -> Bool = FreeLibrary
    let unloadHookAddr = unsafeBitCast(lpfnFreeLibrary, to: LPTHREAD_START_ROUTINE.self)
```
causes a runtime cast failure

```
Swift/Builtin.swift:95: Fatal error: Can't unsafeBitCast between types of different sizes
Current stack trace:
0    (null)                             0x00007fffb7581100 swift_stdlib_reportFatalErrorInFile + 132
```

I'm not sure why this one doesn't work when the `LoadLibraryW` one worked

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,38 +237,277 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+                                       to: LPTHREAD_START_ROUTINE.self)
+    let hUnloadThread = CreateRemoteThread(
+        self.process, nil, 0, unloadHookAddr,
+        unsafeBitCast(hDllModule, to:LPVOID.self), 0, nil)
```
```suggestion
        unsafeBitCast(hDllModule, to: LPVOID.self), 0, nil)
```

hjyamauchi replied:
Done

hjyamauchi commented on tools/swift-inspect/Sources/SwiftInspectClient/SwiftInspectClient.cpp:
```diff
@@ -0,0 +1,218 @@
+  HANDLE Handle;
+  explicit ScopedHandle(HANDLE Handle) noexcept : Handle(Handle) {}
+  ~ScopedHandle() noexcept {
+    if (Handle != NULL) {
```
As I understand, `INVALID_HANDLE_VALUE` != `NULL` and the APIs that we use here returns NULL on failures.

compnerd replied:
Correct, the values are different.  I'm worried that we might try to do a `CloseHandle` on `INVALID_HANDLE_VALUE` as that is not being checked.

hjyamauchi replied:
Done

hjyamauchi commented on tools/swift-inspect/Sources/SwiftInspectClient/SwiftInspectClient.cpp:
```diff
@@ -0,0 +1,218 @@
+    while (HeapWalk(Heap, &Entry)) {
+      if ((!(Entry.wFlags & PROCESS_HEAP_REGION)) &&
+          (!(Entry.wFlags & PROCESS_HEAP_UNCOMMITTED_RANGE)) &&
+          (Entry.wFlags & PROCESS_HEAP_ENTRY_BUSY)) {
```
Done

hjyamauchi commented on tools/swift-inspect/Sources/SwiftInspectClient/SwiftInspectClient.cpp:
```diff
@@ -0,0 +1,218 @@
+      if ((!(Entry.wFlags & PROCESS_HEAP_REGION)) &&
+          (!(Entry.wFlags & PROCESS_HEAP_UNCOMMITTED_RANGE)) &&
+          (Entry.wFlags & PROCESS_HEAP_ENTRY_BUSY)) {
+        if (Count < BUF_NUM_ENTRIES) {
```
Done

hjyamauchi commented on tools/swift-inspect/Sources/SwiftInspectClient/SwiftInspectClient.cpp:
```diff
@@ -0,0 +1,218 @@
+    }
+
+    // Write the remaining entries.
+    if (!SetEvent(ReadEvent.get())) {
```
Done

hjyamauchi commented on tools/swift-inspect/Sources/SwiftInspectClient/SwiftInspectClient.cpp:
```diff
@@ -0,0 +1,218 @@
+BOOL APIENTRY DllMain(HANDLE hModule, DWORD ul_reason_for_call,
+                      LPVOID lpReserved) {
+  if (ul_reason_for_call == DLL_PROCESS_ATTACH) {
+    heapWalk();
```
Yeah, depends on what we need.

hjyamauchi commented on tools/swift-inspect/Sources/swift-inspect/WinSDK+Extentions.swift:
```diff
@@ -0,0 +1,21 @@
+                                    SECTION_MAP_WRITE | SECTION_MAP_READ |
+                                    SECTION_MAP_EXECUTE|SECTION_EXTEND_SIZE
+
+#endif
```
Done

hjyamauchi commented on tools/swift-inspect/Sources/swift-inspect/WinSDK+Extentions.swift:
```diff
@@ -0,0 +1,21 @@
+
+ internal let FILE_MAP_ALL_ACCESS = STANDARD_RIGHTS_REQUIRED | SECTION_QUERY |
+                                    SECTION_MAP_WRITE | SECTION_MAP_READ |
+                                    SECTION_MAP_EXECUTE|SECTION_EXTEND_SIZE
```
Done

hjyamauchi commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,38 +237,277 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+
+    // Set up the shared memory
+    let hMapFile = CreateFileMappingA(INVALID_HANDLE_VALUE,
+                                      UnsafeMutablePointer(bitPattern: 0),
```
Done

hjyamauchi commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,38 +237,277 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+                                      DWORD(PAGE_READWRITE),
+                                      0,
+                                      DWORD(bufSize),
+                                      sharedMemoryName);
```
Done

hjyamauchi commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,38 +237,277 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+                                    DWORD(FILE_MAP_ALL_ACCESS),
+                                    0,
+                                    0,
+                                    SIZE_T(bufSize));
```
Done

hjyamauchi commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,38 +237,277 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+    }
+    defer { CloseHandle(hMapFile) }
+    let buf: LPVOID = MapViewOfFile(hMapFile,
+                                    DWORD(FILE_MAP_ALL_ACCESS),
```
Done

hjyamauchi commented:
Addressed the comments except for the embedding the dll into the executable thing. I imagine that we need to write a windows program that adds the dll as a resource into the executable and run it as part of the `swift-inspect` build process. How do I do that? Any examples?

compnerd commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,38 +255,299 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+    }
+
+    // Unload the dll from the remote process
+    let unloadHookAddr = unsafeBitCast(
```
Is this correct?  This is using the address of `Kernel32.FreeLibrary` in the `swift-inspect` process but invoking it in the remote process, which is not guaranteed to be loaded at the same address right?  I don't think that this is safe to do.  We need to have the function be bound in the process.  The simplest way to do this is to have a secondary entry point that we can invoke in the injected module that can unload itself.

hjyamauchi replied:
I think that kernel32 is special and guaranteed to be loaded at the same address across processes.

https://learn.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-entry-point-function

> Because Kernel32.dll is guaranteed to be loaded in the process address space when the entry-point function is called...

compnerd replied:
I'm not sure I read it the same way - it is that it is loaded before the entry point executes, it doesn't guarantee that the module is loaded at the same address, ASLR can relocate the base address.

hjyamauchi replied:
Yeah, thatwas a misquote, sorry.

I'm reading that kernel32 is loaded at the same address across processes, however. For example, http://www.nynaeve.net/?p=198

Isn't that assumed and necessary when we call `LoadLibraryW` earlier?

compnerd replied:
Yes, that is a mistake as well.  Discussed this offline, and we will use the toolhelp32 to snapshot and find the `HMODULE` associated with kernel32.dll in the image and use `GetProcAddress` to get the proper location.

compnerd reviewed (commented):
This is nearly ready!  I think that the question around the invocation of `FreeLibrary` is something we need to think through, but the rest is looking really good!

tristanlabelle commented on tools/swift-inspect/Sources/SwiftInspectClient/SwiftInspectClient.cpp:
```diff
@@ -0,0 +1,221 @@
+
+  // Collect heaps. This is a loop because GetProcessHeaps requires
+  // specifying the max number of heaps to get upfront.
+  std::vector<HANDLE> Heaps;
```
Yes! C++ standard library types have default constructors invoked implicitly and won't be uninitialized.

hjyamauchi commented on tools/swift-inspect/Sources/SwiftInspectClient/SwiftInspectClient.cpp:
```diff
@@ -0,0 +1,221 @@
+        continue;
+      }
+      assert(Count < BUF_NUM_ENTRIES);
+      Buf.as<HeapEntry>()[Count].Address =
```
Done

hjyamauchi commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -60,11 +62,13 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+      if !ReadProcessMemory(
+        process.process, LPVOID(bitPattern: UInt(address)),
+        buffer, size, nil)
+      {
```
I applied swift-format, not swiftformat, with the default setting except
```
{
  "indentConditionalCompilationBlocks": false,
}
```
and this is what it gives :)

<img width="1288" alt="Screenshot 2023-07-13 at 5 15 56 PM" src="https://github.com/apple/swift/assets/56735936/e875ea70-cd71-496e-af0c-6909ea2776d4">

hjyamauchi commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,38 +255,299 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+    // The size of the shared memory buffer and the names of shared
+    // memory and event objects
+    let bufSize = Int(BUF_SIZE)
+    let sharedMemoryName = SHARED_MEM_NAME_PREFIX + "-" + String(dwProcessId)
```
Done

hjyamauchi commented on tools/swift-inspect/Sources/swift-inspect/WindowsRemoteProcess.swift:
```diff
@@ -233,38 +255,299 @@ internal final class WindowsRemoteProcess: RemoteProcess {
+    var hKernel32 = HMODULE(bitPattern: 0)
+    "kernel32".withCString {
+      hKernel32 = GetModuleHandleA($0)
+    }
```
Done

hjyamauchi commented:
PTAL

hjyamauchi commented:
Updated to get the remote `LoadLibraryW/FreeLibrary` addresses from the module snapshot and `GetProcAddress()` and no to rely on them being at the same addresses across processes.

compnerd commented:
@swift-ci please test
